beacon-chain/core/peerdas/reconstruction.go

```diff
@@ -2,75 +2,321 @@ package peerdas
+		cells, err := kzg.ComputeCells(&kzgBlob)
+		if err != nil {
+			return nil, errors.Wrap(err, "compute cells")
+		}
+
+		var proofs []kzg.Proof
+		for idx := uint64(i) * numberOfColumns; idx < (uint64(i)+1)*numberOfColumns; idx++ {
+			var kzgProof kzg.Proof
+			if copy(kzgProof[:], cellProofs[idx]) != len(kzgProof) {
+				return nil, errors.New("wrong KZG proof size - should never happen")
+			}
+
+			proofs = append(proofs, kzgProof)
+		}
+
+		cellsProofs := kzg.CellsAndProofs{Cells: cells, Proofs: proofs}
+		cellsAndProofs = append(cellsAndProofs, cellsProofs)
+	}
+
+	dataColumnSidecars, err := DataColumnSidecars(block, cellsAndProofs)
+	if err != nil {
+		return nil, errors.Wrap(err, "data column sidcars")
+	}
+
+	return dataColumnSidecars, nil
+}
+
+// ReconstructBlobs constructs verified read only blobs sidecars from verified read only blob sidecars.
+// The following constraints must be satisfied:
+//   - All `dataColumnSidecars` has to be committed to the same block, and
+//   - `dataColumnSidecars` must be sorted by index and should not contain duplicates.
+//   - `dataColumnSidecars` must contain either all sidecars corresponding to (non-extended) blobs,
+//     or either enough sidecars to reconstruct the blobs.
+func ReconstructBlobs(block blocks.ROBlock, verifiedDataColumnSidecars []blocks.VerifiedRODataColumn, indices []int) ([]*blocks.VerifiedROBlob, error) {
+	// Return early if no blobs are requested.
+	if len(indices) == 0 {
+		return nil, nil
+	}
+
+	if len(verifiedDataColumnSidecars) == 0 {
+		return nil, ErrNotEnoughDataColumnSidecars
+	}
+
+	// Check if the sidecars are sorted by index and do not contain duplicates.
+	previousColumnIndex := verifiedDataColumnSidecars[0].Index
+	for _, dataColumnSidecar := range verifiedDataColumnSidecars[1:] {
+		columnIndex := dataColumnSidecar.Index
+		if columnIndex <= previousColumnIndex {
+			return nil, ErrDataColumnSidecarsNotSortedByIndex
+		}
+
+		previousColumnIndex = columnIndex
+	}
+
+	// Check if we have enough columns.
+	cellsPerBlob := fieldparams.CellsPerBlob
+	if len(verifiedDataColumnSidecars) < cellsPerBlob {
+		return nil, ErrNotEnoughDataColumnSidecars
+	}
+
+	// Check if the blob index is too high.
+	commitments, err := block.Block().Body().BlobKzgCommitments()
+	if err != nil {
+		return nil, errors.Wrap(err, "blob KZG commitments")
+	}
+
+	for _, blobIndex := range indices {
+		if blobIndex >= len(commitments) {
+			return nil, ErrBlobIndexTooHigh
+		}
+	}
+
+	// Check if the data column sidecars are aligned with the block.
+	dataColumnSidecars := make([]blocks.RODataColumn, 0, len(verifiedDataColumnSidecars))
+	for _, verifiedDataColumnSidecar := range verifiedDataColumnSidecars {
+		dataColumnSicecar := verifiedDataColumnSidecar.RODataColumn
+		dataColumnSidecars = append(dataColumnSidecars, dataColumnSicecar)
+	}
+
+	if err := DataColumnsAlignWithBlock(block, dataColumnSidecars); err != nil {
+		return nil, errors.Wrap(err, "data columns align with block")
+	}
```

Comment:
why do we need this check if we assume the data column sidecars are already trusted and verified judging by the name `verifiedDataColumnSidecars`

Reply:
`DataColumnsAlignWithBlock` does not run any verification check as defined in [p2p-interface.md](https://github.com/ethereum/consensus-specs/blob/dev/specs/fulu/p2p-interface.md).

`DataColumnsAlignWithBlock` only ensures all `dataColumnSidecars` correspond to `block`.
Because `block` and `dataColumns` are arguments of `ReconstructBlobs`, the caller can call this function with any block/sidecars, including non related block/sidecars.

Reply:
Im sorry, I still dont understand why we need `DataColumnsAlignWithBlock` in `ReconstructBlobs`
Callers should already verified or trusting the input data is correct
Who are the attack vectors? Are we worry about EL is malicious here?
Similar when we call `GetBlobs` today, we dont verify the blobs that received over EL is correct

Reply:
I am not strongly against removing this check.
However, it report the responsibility of checking that block and data column sidecars are effectively aligned on the caller. Failing to do so may make the `ReconstructBlobs` function to panic (because, for example, if the commitment count is not equal to the cells count of a column).

Currently, this `ReconstructBlobs` is only used in the handler of the blob sidecars beacon API.
So I can remove the alignment check in the function, and add a warning in the godoc of `ReconstructBlobs` saying that the function may panic if `block` and `verifiedRoDataColumnSidecars` are not aligned.

Reply:
As a general approach, I like when a function never panics, even if its inputs are garbabe.

Reply:
I'm not strongly pushing to remove the check. It was more of a question to clarify my understanding, and it really depends on who the caller is.
If the caller is from internal, trusted Prysm services or execution layer clients (after `getBlobs`), then I don't think the check is necessary.
But if the caller is coming from the beacon API and users are providing the arguments, then I think it's good to keep the check.